Intermittent test failures

<!--
    NOTE: This template is for use by maintainers only. Please do not submit
    an issue using this template unless you have been specifically asked to
    do so.
-->
### Proposed Changes
We should adjust this test to not rely on random chance, which causes [intermittent failures](https://github.com/nautobot/nautobot/actions/runs/23448715234/job/68220681975):
```python
======================================================================
FAIL: test_cleanup_object_changes (nautobot.core.tests.test_jobs.LogsCleanupTestCase)
With unconstrained permissions, all ObjectChanges before the cutoff should be deleted.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.20/x64/lib/python3.10/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/opt/hostedtoolcache/Python/3.10.20/x64/lib/python3.10/unittest/case.py", line 591, in run
    self._callTestMethod(testMethod)
  File "/opt/hostedtoolcache/Python/3.10.20/x64/lib/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "/home/runner/work/nautobot/nautobot/nautobot/core/tests/test_jobs.py", line 734, in test_cleanup_object_changes
    self.assertTrue(ObjectChange.objects.filter(time__gte=cutoff).exists())
  File "/opt/hostedtoolcache/Python/3.10.20/x64/lib/python3.10/unittest/case.py", line 687, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true

----------------------------------------------------------------------
```

<!-- Provide justification for the proposed change(s). -->
### Justification
QOL

--- a/changes/8734.housekeeping
+++ b/changes/8734.housekeeping
@@ -1,1 +1,1 @@
-Changed ObjectChangeFactory to use `factory_boy` instead of `faker` for deterministic timestamps for the `time` field for tests.
+Changed ObjectChangeFactory to use set start and end dates for deterministic timestamps for the `time` field for tests.

--- a/changes/8734.housekeeping
+++ b/changes/8734.housekeeping
@@ -1,0 +1,1 @@
+Changed test_cleanup_object_changes to ensure that it always has something to clean up and something to find after cleanup.

--- a/changes/8734.housekeeping
+++ b/changes/8734.housekeeping
@@ -1,1 +1,1 @@
-Changed test_cleanup_object_changes to ensure that it always has something to clean up and something to find after cleanup.
+Changed ObjectChangeFactory to use `factory_boy` instead of `faker` for deterministic timestamps for the `time` field for tests.

--- a/nautobot/core/tests/test_jobs.py
+++ b/nautobot/core/tests/test_jobs.py
@@ -722,10 +722,6 @@ class LogsCleanupTestCase(TransactionTestCase):
     def test_cleanup_object_changes(self):
         """With unconstrained permissions, all ObjectChanges before the cutoff should be deleted."""
         cutoff = timezone.now() - timedelta(days=60)
-        # Pin one record inside the window so the post-cleanup assertTrue has something to find.
-        ObjectChangeFactory(time=timezone.now() - timedelta(days=30))
-        # Pin one record outside the window so the assertFalse actually proves deletion occurred.
-        ObjectChangeFactory(time=timezone.now() - timedelta(days=90))
         create_job_result_and_run_job(
             "nautobot.core.jobs.cleanup",
             "LogsCleanup",

--- a/nautobot/core/tests/test_jobs.py
+++ b/nautobot/core/tests/test_jobs.py
@@ -9,6 +9,7 @@ from django.contrib.contenttypes.models import ContentType
 from django.core.cache import cache
 from django.core.files.base import ContentFile
 from django.utils import timezone
+import time_machine
 import yaml
 
 from nautobot.circuits.models import Circuit, CircuitType, Provider
@@ -675,63 +676,67 @@ class LogsCleanupTestCase(TransactionTestCase):
     )
     def test_cleanup_job_results(self):
         """With unconstrained permissions, all JobResults before the cutoff should be deleted."""
-        cutoff = timezone.now() - timedelta(days=60)
-        job_results_to_be_deleted = JobResult.objects.filter(date_done__lt=cutoff)
-        job_results_to_be_deleted_count = job_results_to_be_deleted.count()
-        job_log_entry_to_be_deleted_count = JobLogEntry.objects.filter(job_result__in=job_results_to_be_deleted).count()
-        objectmetadata_to_be_deleted_count = ObjectMetadata.objects.filter(
-            assigned_object_id__in=job_results_to_be_deleted,
-            assigned_object_type=ContentType.objects.get_for_model(JobResult),
-        ).count()
+        with time_machine.travel("2024-10-01 00:00 +0000"):
+            cutoff = timezone.now() - timedelta(days=60)
+            job_results_to_be_deleted = JobResult.objects.filter(date_done__lt=cutoff)
+            job_results_to_be_deleted_count = job_results_to_be_deleted.count()
+            job_log_entry_to_be_deleted_count = JobLogEntry.objects.filter(
+                job_result__in=job_results_to_be_deleted
+            ).count()
+            objectmetadata_to_be_deleted_count = ObjectMetadata.objects.filter(
+                assigned_object_id__in=job_results_to_be_deleted,
+                assigned_object_type=ContentType.objects.get_for_model(JobResult),
+            ).count()
 
-        with self.assertLogs("nautobot.events") as cm:
-            job_result = create_job_result_and_run_job(
-                "nautobot.core.jobs.cleanup",
-                "LogsCleanup",
-                cleanup_types=[CleanupTypes.JOB_RESULT],
-                max_age=60,
+            with self.assertLogs("nautobot.events") as cm:
+                job_result = create_job_result_and_run_job(
+                    "nautobot.core.jobs.cleanup",
+                    "LogsCleanup",
+                    cleanup_types=[CleanupTypes.JOB_RESULT],
+                    max_age=60,
+                )
+            self.assertFalse(JobResult.objects.filter(date_done__lt=cutoff).exists(), cm.output)
+            self.assertTrue(JobResult.objects.filter(date_done__gte=cutoff).exists(), cm.output)
+            self.assertTrue(ObjectChange.objects.filter(time__lt=cutoff).exists(), cm.output)
+            self.assertTrue(ObjectChange.objects.filter(time__gte=cutoff).exists(), cm.output)
+
+            started_logs = {
+                "job_result_id": str(job_result.id),
+                "job_name": "Logs Cleanup",
+                "user_name": job_result.user.username,
+                "job_kwargs": {"cleanup_types": ["extras.JobResult"], "max_age": 60},
+            }
+            self.assertEqual(
+                cm.output[0],
+                f"INFO:nautobot.events.nautobot.jobs.job.started:{json.dumps(started_logs, indent=4)}",
             )
-        self.assertFalse(JobResult.objects.filter(date_done__lt=cutoff).exists(), cm.output)
-        self.assertTrue(JobResult.objects.filter(date_done__gte=cutoff).exists(), cm.output)
-        self.assertTrue(ObjectChange.objects.filter(time__lt=cutoff).exists(), cm.output)
-        self.assertTrue(ObjectChange.objects.filter(time__gte=cutoff).exists(), cm.output)
 
-        started_logs = {
-            "job_result_id": str(job_result.id),
-            "job_name": "Logs Cleanup",
-            "user_name": job_result.user.username,
-            "job_kwargs": {"cleanup_types": ["extras.JobResult"], "max_age": 60},
-        }
-        self.assertEqual(
-            cm.output[0],
-            f"INFO:nautobot.events.nautobot.jobs.job.started:{json.dumps(started_logs, indent=4)}",
-        )
+            started_logs["job_output"] = {
+                "extras.JobResult": job_results_to_be_deleted_count,
+                "extras.JobLogEntry": job_log_entry_to_be_deleted_count,
+            }
+            if objectmetadata_to_be_deleted_count > 0:
+                started_logs["job_output"]["extras.ObjectMetadata"] = objectmetadata_to_be_deleted_count
 
-        started_logs["job_output"] = {
-            "extras.JobResult": job_results_to_be_deleted_count,
-            "extras.JobLogEntry": job_log_entry_to_be_deleted_count,
-        }
-        if objectmetadata_to_be_deleted_count > 0:
-            started_logs["job_output"]["extras.ObjectMetadata"] = objectmetadata_to_be_deleted_count
-
-        self.assertEqual(
-            cm.output[1],
-            f"INFO:nautobot.events.nautobot.jobs.job.completed:{json.dumps(started_logs, indent=4)}",
-        )
+            self.assertEqual(
+                cm.output[1],
+                f"INFO:nautobot.events.nautobot.jobs.job.completed:{json.dumps(started_logs, indent=4)}",
+            )
 
     def test_cleanup_object_changes(self):
         """With unconstrained permissions, all ObjectChanges before the cutoff should be deleted."""
-        cutoff = timezone.now() - timedelta(days=60)
-        create_job_result_and_run_job(
-            "nautobot.core.jobs.cleanup",
-            "LogsCleanup",
-            cleanup_types=[CleanupTypes.OBJECT_CHANGE],
-            max_age=60,
-        )
-        self.assertTrue(JobResult.objects.filter(date_done__lt=cutoff).exists())
-        self.assertTrue(JobResult.objects.filter(date_done__gte=cutoff).exists())
-        self.assertFalse(ObjectChange.objects.filter(time__lt=cutoff).exists())
-        self.assertTrue(ObjectChange.objects.filter(time__gte=cutoff).exists())
+        with time_machine.travel("2024-10-01 00:00 +0000"):
+            cutoff = timezone.now() - timedelta(days=60)
+            create_job_result_and_run_job(
+                "nautobot.core.jobs.cleanup",
+                "LogsCleanup",
+                cleanup_types=[CleanupTypes.OBJECT_CHANGE],
+                max_age=60,
+            )
+            self.assertTrue(JobResult.objects.filter(date_done__lt=cutoff).exists())
+            self.assertTrue(JobResult.objects.filter(date_done__gte=cutoff).exists())
+            self.assertFalse(ObjectChange.objects.filter(time__lt=cutoff).exists())
+            self.assertTrue(ObjectChange.objects.filter(time__gte=cutoff).exists())
 
 
 class BulkEditTestCase(TransactionTestCase):

--- a/nautobot/core/tests/test_jobs.py
+++ b/nautobot/core/tests/test_jobs.py
@@ -722,6 +722,10 @@ class LogsCleanupTestCase(TransactionTestCase):
     def test_cleanup_object_changes(self):
         """With unconstrained permissions, all ObjectChanges before the cutoff should be deleted."""
         cutoff = timezone.now() - timedelta(days=60)
+        # Pin one record inside the window so the post-cleanup assertTrue has something to find.
+        ObjectChangeFactory(time=timezone.now() - timedelta(days=30))
+        # Pin one record outside the window so the assertFalse actually proves deletion occurred.
+        ObjectChangeFactory(time=timezone.now() - timedelta(days=90))
         create_job_result_and_run_job(
             "nautobot.core.jobs.cleanup",
             "LogsCleanup",

--- a/nautobot/extras/factory.py
+++ b/nautobot/extras/factory.py
@@ -218,15 +218,15 @@ class JobResultFactory(BaseModelFactory):
             return
         if extracted:
             return
-        # Create a date_created in the past, but not too far in the past
-        self.date_created = faker.Faker().date_time_between(start_date="-1y", end_date="-1w", tzinfo=timezone.utc)
-        self.date_started = faker.Faker().date_time_between(
-            start_date=self.date_created, end_date="-1d", tzinfo=timezone.utc
+        # Fixed absolute dates with factory_boy's seeded random for full determinism.
+        # Avoids standalone faker (bypasses seed) and relative dates (faker#2149).
+        start = datetime(2024, 1, 1, tzinfo=timezone.utc)
+        end = datetime(2025, 1, 1, tzinfo=timezone.utc)
+        seconds_range = int((end - start).total_seconds())
+        timestamps = sorted(
+            start + timedelta(seconds=factory.random.randgen.randint(0, seconds_range)) for _ in range(3)
         )
-        # TODO, should we create "in progress" job results without a date_done value as well?
-        self.date_done = faker.Faker().date_time_between(
-            start_date=self.date_started, end_date="now", tzinfo=timezone.utc
-        )
+        self.date_created, self.date_started, self.date_done = timestamps
 
 
 class MetadataChoiceFactory(BaseModelFactory):
@@ -474,13 +474,10 @@ class ObjectChangeFactory(BaseModelFactory):
             if extracted:
                 self.time = extracted
             else:
-                # Generate timestamps between 30 and 90 days ago using factory_boy's seeded
-                # random. This ensures records land on both sides of the 60-day cleanup cutoff.
-                # Standalone faker.Faker() bypasses factory_boy's seed, and relative date
-                # strings like "-1y" are non-deterministic (faker#2149).
-                now = datetime.now(tz=timezone.utc)
-                start = now - timedelta(days=90)
-                end = now - timedelta(days=30)
+                # Fixed absolute dates with factory_boy's seeded random for full determinism.
+                # Avoids standalone faker (bypasses seed) and relative dates (faker#2149).
+                start = datetime(2024, 1, 1, tzinfo=timezone.utc)
+                end = datetime(2025, 1, 1, tzinfo=timezone.utc)
                 seconds_range = int((end - start).total_seconds())
                 self.time = start + timedelta(seconds=factory.random.randgen.randint(0, seconds_range))
 

--- a/nautobot/extras/factory.py
+++ b/nautobot/extras/factory.py
@@ -474,12 +474,15 @@ class ObjectChangeFactory(BaseModelFactory):
             if extracted:
                 self.time = extracted
             else:
-                # Use factory_boy's seeded random (not standalone faker) with absolute dates
-                # to ensure deterministic output. Standalone faker.Faker() bypasses factory_boy's
-                # seed, and relative date strings like "-1y" are non-deterministic (faker#2149).
+                # Generate timestamps between 30 and 90 days ago using factory_boy's seeded
+                # random. This ensures records land on both sides of the 60-day cleanup cutoff.
+                # Standalone faker.Faker() bypasses factory_boy's seed, and relative date
+                # strings like "-1y" are non-deterministic (faker#2149).
                 now = datetime.now(tz=timezone.utc)
-                seconds_in_year = 365 * 24 * 60 * 60
-                self.time = now - timedelta(seconds=factory.random.randgen.randint(0, seconds_in_year))
+                start = now - timedelta(days=90)
+                end = now - timedelta(days=30)
+                seconds_range = int((end - start).total_seconds())
+                self.time = start + timedelta(seconds=factory.random.randgen.randint(0, seconds_range))
 
 
 class RoleFactory(OrganizationalModelFactory):

--- a/nautobot/extras/factory.py
+++ b/nautobot/extras/factory.py
@@ -1,4 +1,4 @@
-from datetime import timezone
+from datetime import datetime, timedelta, timezone
 import json
 
 from django.contrib.auth import get_user_model
@@ -474,7 +474,12 @@ class ObjectChangeFactory(BaseModelFactory):
             if extracted:
                 self.time = extracted
             else:
-                self.time = faker.Faker().date_time_between(start_date="-1y", end_date="now", tzinfo=timezone.utc)
+                # Use factory_boy's seeded random (not standalone faker) with absolute dates
+                # to ensure deterministic output. Standalone faker.Faker() bypasses factory_boy's
+                # seed, and relative date strings like "-1y" are non-deterministic (faker#2149).
+                now = datetime.now(tz=timezone.utc)
+                seconds_in_year = 365 * 24 * 60 * 60
+                self.time = now - timedelta(seconds=factory.random.randgen.randint(0, seconds_in_year))
 
 
 class RoleFactory(OrganizationalModelFactory):

--- a/nautobot/extras/tests/test_changelog.py
+++ b/nautobot/extras/tests/test_changelog.py
@@ -458,7 +458,7 @@ class ChangeLogAPITest(APITestCase):
         resp = execute_query(gql_payload, user=self.user)
         self.assertIsNone(resp.errors)
         self.assertIsInstance(resp.data.get("query"), list)
-        # ObjectChangeFactory creates records in the last year only; there shouldn't be any in this filtered response.
+        # ObjectChangeFactory creates records between 30 and 90 days ago; there shouldn't be any in this filtered response.
         self.assertEqual(len(resp.data.get("query")), 0)
 
     @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"])

--- a/nautobot/extras/tests/test_changelog.py
+++ b/nautobot/extras/tests/test_changelog.py
@@ -458,7 +458,7 @@ class ChangeLogAPITest(APITestCase):
         resp = execute_query(gql_payload, user=self.user)
         self.assertIsNone(resp.errors)
         self.assertIsInstance(resp.data.get("query"), list)
-        # ObjectChangeFactory creates records between 30 and 90 days ago; there shouldn't be any in this filtered response.
+        # ObjectChangeFactory creates records with fixed dates in 2024; there shouldn't be any in this filtered response.
         self.assertEqual(len(resp.data.get("query")), 0)
 
     @override_settings(EXEMPT_VIEW_PERMISSIONS=["*"])

